packages/core/src/PasteRule.ts

```diff
@@ -212,45 +261,45 @@ export function pasteRulesPlugin(props: { editor: Editor; rules: PasteRule[] }):
         const isPaste = transaction.getMeta('uiEvent') === 'paste' && !isPastedFromProseMirror
         const isDrop = transaction.getMeta('uiEvent') === 'drop' && !isDroppedFromProseMirror
 
-        if (!isPaste && !isDrop) {
+        // if PasteRule is triggered by insertContent()
+        const simulatedPasteMeta = transaction.getMeta('applyPasteRules')
+        const isSimulatedPaste = !!simulatedPasteMeta
+
+        if (!isPaste && !isDrop && !isSimulatedPaste) {
           return
         }
 
-        // stop if there is no changed range
+        // Handle simulated paste
+        if (isSimulatedPaste) {
+          const { from, text } = simulatedPasteMeta
+          const to = from + text.length
+          const pasteEvt = createClipboardPasteEvent(text)
+
+          return processEvent({
+            rule,
+            state,
+            from,
+            to: { b: to },
+            pasteEvt,
+          })
+        }
+
+        // handle actual paste/drop
         const from = oldState.doc.content.findDiffStart(state.doc.content)
         const to = oldState.doc.content.findDiffEnd(state.doc.content)
 
+        // stop if there is no changed range
         if (!isNumber(from) || !to || from === to.b) {
           return
         }
 
-        // build a chainable state
-        // so we can use a single transaction for all paste rules
-        const tr = state.tr
-        const chainableState = createChainableState({
-          state,
-          transaction: tr,
-        })
-
-        const handler = run({
-          editor,
-          state: chainableState,
-          from: Math.max(from - 1, 0),
-          to: to.b - 1,
+        return processEvent({
```

Comment:
I see why this `processEvent` function was extracted, but all that it is doing is switching the arguments that it is operating on.
I think it'd be clearer to inline it again, and have the `if (isSimulatedPaste) {` switch the values it operates on.

Reply:
Haha, tbh that's what I had done before and I thought it might be easier to understand that way 😂